stack/maxima/contrib/linearalgebra.mac

```diff
@@ -0,0 +1,1205 @@
+ * The vectors themselves may be different objects. Supported objects include:
+ * * 1xN or Mx1 matrices
+ * * Lists
+ * * Vectors using c or r notation
+ * * Sets, ntuples
+ * Other inert functions like sequence should work too, but 
+ * The collection may be a list, set, ntuple, span, or matrix
+ * If given a matrix, then returns the columns as a list, with each column also being a list.
+ *
+ * @param[] ex A list, set, ntuple or span containing vectors as matrices, lists, c, r, sets or ntuples, or a matrix
+ * @return[list] A list of lists
+ */
+make_list_of_lists(ex):= block([op1],
+  op1: safe_op(ex),
+  if not(member(op1,["[","ntuple","{","span","matrix"])) then return(ex),
+  ex: vec_convert(ex),
+  if vectorp(ex) then return([list_matrix_entries(ex)]),
+  if is(op1="matrix") then return(args(transpose(ex))),
+  ex: args(ex),
+  if ev(every(lambda([ex2],constantp(ex2) or atom(ex2)),ex),simp) then return([ex]),
+  ex: map(lambda([ex2],if vectorp(ex2) then list_matrix_entries(ex2) else args(ex2)),ex),
+  return(ex)
+);
+
+/**
+ * Given a list of lists, return the matrix that has those sub-lists as columns.
+ * Naming convention from numpy
+ * If the sub-lists do not conform then returns the list of lists instead
+ * Note: row_stack is absent because the remaining functions that accept matrices assume that we are considering columns.
+ *
+ * @param[list] ex a list of lists. Each sub-list must be the same length
+ * @return[matrix] A matrix whose columns are the entries of ex
+ */
+column_stack(ex):= block([ex2],
+  ex2: errcatch(transpose(apply(matrix,args(ex)))),
+  if emptyp(ex2) then return(ex2) else return(first(ex2))
+);
+
+/* TODO function to convert list of lists to list of column vectors. */
+
+/*********************************************************************************/
+/* Comparison functions                                                          */
+/*********************************************************************************/
+
+/**
+ * Given a list of lists or a matrix, determine whether the list elements or columns are linearly independent.
+ * If ex is a matrix, it checks for full column rank, not row rank. To check for full rank generally, use invertiblep.
+ * 
+ * @param[list or matrix] ex Either a list of lists (use make_list_of_lists if needed) or a matrix
+ * @return[boolean] Is the collection of vectors linearly independent?
+ */
+lin_indp(ex):= block(
+  if matrixp(ex) then return(is(rank(ex) = ev(second(matrix_size(ex)),simp)))
+  else ex: column_stack(ex),
+  if matrixp(ex) then return(is(rank(ex) = ev(second(matrix_size(ex)),simp))),
+  return(false)
+);
+
+/**
+ * Given a pair of matrices, check whether they are row equivalent.
+ * i.e. can one matrix be transformed into the other with elementary row operations?
+ *
+ * @param[matrix] ex a mxn matrix
+ * @param[matrix] ta a mxn matrix
+ * @return[boolean] Are ex and ta row equivalent?
+ */
+row_equiv(ex,ta):= block(
```

Comment:
Should this be called `row_equivp`, given that this is a predicate function?

Same for `col_equiv` and `subspace_equiv`.

Reply:
You're probably correct here. For some reason I had in my head that a predicate function would only take a single input, but a quick Google disproves that. I am wanting to do a (hopefully) final pass of function name changes before starting on question-writing in earnest, so I will fix these up then. My colleague had some other suggestions that I want to implement too.